whatweb.1
.TH WHATWEB 1 "December 14th, 2020"
.SH NAME
WhatWeb \- Next generation Web scanner. Identify technologies used by websites.
.SH SYNOPSIS
.B whatweb
.RI [ options ] 
.RI <URLs>

.br
.SH DESCRIPTION
.PP
WhatWeb identifies websites. It's goal is to answer the question, "What is that Website?". WhatWeb recognises web technologies including content management systems (CMS), blogging platforms, statistic/analytics packages, JavaScript libraries, web servers, and embedded devices. WhatWeb has over 1800 plugins, each to recognise something different. WhatWeb also identifies version numbers, email addresses, account ID's, web framework modules, SQL errors, and more.

WhatWeb can be stealthy and fast, or thorough but slow. WhatWeb supports an aggression level to control the trade off between speed and reliability. When you visit a website in your browser, the transaction includes many hints of what web technologies are powering that website. Sometimes a single webpage visit contains enough information to identify a website but when it does not, WhatWeb can interrogate the website further. The default level of aggression, called 'passive', is the fastest and requires only one HTTP request of a website. This is suitable for scanning public websites. More aggressive modes were developed for in penetration tests.

Most WhatWeb plugins are thorough and recognise a range of cues from subtle to obvious. For example, most WordPress websites can be identified by the meta HTML tag, e.g. '<meta name="generator" content="WordPress 2.6.5">', but a minority of WordPress websites remove this identifying tag but this does not thwart WhatWeb. The WordPress WhatWeb plugin has over 15 tests, which include checking the favicon, default installation files, login pages, and checking for "/wp-content/" within relative links.

.TP
.B
Features:
.TP
	* Over 1800 plugins
.TP
	* Control the trade off between speed/stealth and reliability
.TP
	* Performance tuning. Control how many websites to scan concurrently.
.TP
	* Multiple log formats: Brief (greppable), Verbose (human readable), XML, JSON, MagicTree, RubyObject, MongoDB, SQL.
.TP
	* Proxy support including TOR
.TP
	* Custom HTTP headers
.TP
	* Basic HTTP authentication
.TP
	* Control over webpage redirection
.TP
	* IP address ranges
.TP
	* Fuzzy matching
.TP
	* Result certainty awareness
.TP
	* Custom plugins defined on the command line
.TP
	* IDN (International Domain Name) support


.SH TARGET SELECTION
.TP
.B \<TARGETs\>
Enter URLs, hostnames, IP adddresses, filenames or IP ranges in CIDR, x.x.x-x, or x.x.x.x-x.x.x.x format.
.TP
.B \-\-input-file=FILE \-i
Identify URLs found in FILE
.SH TARGET MODIFICATION
.TP
.B \-\-url-prefix
Add a prefix to target URLs
.TP
.B \-\-url-suffix
Add a suffix to target URLs
.TP
.B \-\-url-pattern
Insert the targets into a URL. Requires --input-file,
eg. www.example.com/%insert%/robots.txt
.SH AGGRESSION
.TP
The aggression level controls the trade-off between speed/stealth and reliability.
.TP
.B \-\-aggression \-a=LEVEL
Set the aggression level. Default: 1.
.TP .nf
 1. Stealthy       Makes one HTTP request per target and also follows redirects. 
 3. Aggressive     If a level 1 plugin is matched, additional requests will be made.
 4. Heavy          Makes a lot of HTTP requests per target. URLs from all plugins are attempted.
.ni
.SH HTTP OPTIONS
.TP
.B  \-\-user-agent, \-U=AGENT
Identify as AGENT instead of WhatWeb/0.4.9.
.TP
.B  \-\-header, \-H
Add an HTTP header. eg "Foo:Bar". Specifying a default header will replace it. Specifying an empty value, e.g. "User-Agent:" will remove it.
.TP
.B  \-\-follow\-redirect=WHEN
Control when to follow redirects. WHEN may be `never', `http-only', `meta-only', `same-site', or `always'. Default: always.
.TP
.B  \-\-max\-redirects=NUM
Maximum number of redirects. Default: 10.

.SH AUTHENTICATION
.TP
.B  \-\-user, \-u=<user:password>
HTTP basic authentication.
.TP
.B  \-\-cookie, \-c=COOKIES
Use cookies, e.g. 'name=value; name2=value2'.

.SH PROXY
.B  \-\-proxy
<hostname[:port]> Set proxy hostname and port. Default: 8080.
.TP
.B  \-\-proxy-user
<username:password> Set proxy user and password.

.SH PLUGINS
.TP
.B  \-\-list\-plugins, \-l
List all plugins.
.TP
.B  \-\-info\-plugins, \-I=[SEARCH]
List all plugins with detailed information. Optionally search with keywords in a comma delimited list.
.TP
.B  \-\-search\-plugins=STRING
Search plugins for a keyword.
.TP
.B  \-\-plugins, -p=LIST
Select plugins. LIST is a comma delimited set of selected plugins. Default is all. 
Each element can be a directory, file or plugin name and can optionally have a modifier, +/-.
.TP
.nf
Examples: +/tmp/moo.rb,+/tmp/foo.rb
title,md5,+./plugins-disabled/
./plugins-disabled,-md5
\-p + is a shortcut for -p +plugins\-disabled.
.ni
.TP
.B  \-\-grep, \-g=STRING|REGEXP
Search for STRING or a Regular Expression. Shows only the results that match.
.TP
.nf
Examples: \-\/-grep "hello"
\-\-grep "/he[l]*o/"
.ni
.TP
.B  \-\-custom\-plugin=DEFINITION
Define a custom plugin named Custom-Plugin,
.TP .nf
Examples: ":text=>'powered by abc'"
":version=>/powered[ ]?by ab[0-9]/"
":ghdb=>'intitle:abc \"powered by abc\"'"
":md5=>'8666257030b94d3bdb46e05945f60b42'"
"{:text=>'powered by abc'}"
.ni
.TP
.B  \-\-dorks=PLUGIN
List Google dorks for the selected plugin.


.SH OUTPUT
.TP
.B  \-\-verbose, \-v
Verbose output includes plugin descriptions. Use twice for debugging.
.TP
.B  \-\-colour,\-\-color=WHEN
control whether colour is used. WHEN may be `never', `always', or `auto'.
.TP
.B  \-\-quiet, \-q
Do not display brief logging to STDOUT.
.TP
.B  \-\-no\-errors
Suppress error messages.

.SH LOGGING
.TP
.B  \-\-log\-brief=FILE
Log brief, one\-line output.
.TP
.B  \-\-log\-verbose=FILE
Log verbose output.
.TP
.B  \-\-log\-errors=FILE
Log errors.
.TP
.B  \-\-log\-xml=FILE
Log XML format.
.TP
.B  \-\-log\-json=FILE
Log JSON format.
.TP
.B  \-\-log\-sql=FILE
Log SQL INSERT statements.
.TP
.B  \-\-log\-sql\-create=FILE
Create SQL database tables.
.TP
.B  \-\-log\-json\-verbose=FILE
Log JSON Verbose format.
.TP
.B  \-\-log\-magictree=FILE
Log MagicTree XML format.
.TP
.B  \-\-log\-object=FILE
Log Ruby object inspection format.
.TP
.B  \-\-log\-mongo\-database
Name of the MongoDB database.
.TP
.B  \-\-log\-mongo\-collection
Name of the MongoDB collection. Default: whatweb.
.TP
.B  \-\-log\-mongo\-host
MongoDB hostname or IP address. Default: 0.0.0.0.
.TP
.B  \-\-log\-mongo\-username
MongoDB username. Default: nil.
.TP
.B  \-\-log\-mongo\-password
MongoDB password. Default: nil.
.TP
.B  \-\-log\-elastic\-index
Name of the index to store results. Default: whatweb 
.TP
.B  \-\-log\-elastic\-host
Host:port of the elastic http interface. Default: 127.0.0.1:9200s
  

.SH PERFORMANCE & STABILITY
.TP
.B  \-\-max\-threads, \-t
Number of simultaneous threads. Default: 25.
.TP
.B  \-\-open\-timeout
Time in seconds. Default: 15.
.TP
.B  \-\-read\-timeout
Time in seconds. Default: 30.
.TP
.B  \-\-wait=SECONDS
Wait SECONDS between connections. This is useful when using a single thread.

.SH HELP & MISCELLANEOUS
.TP
.B  \-\-short\-help
Short usage help.
.TP
.B  \-\-help, \-h
Complete usage help.
.TP
.B  \-\-debug
Raise errors in plugins.
.TP
.B  \-\-version
Display version information.

.SH EXAMPLE USAGE
.TP
.B Scan example.com.
 ./whatweb example.com
.TP
.B Scan reddit.com slashdot.org with verbose plugin descriptions.
 ./whatweb -v reddit.com slashdot.org
.TP
.B An aggressive scan of wired.com detects the exact version of WordPress.
 ./whatweb -a 3 www.wired.com
.TP
.B Scan the local network quickly and suppress errors.
 whatweb --no-errors 192.168.0.0/24
.TP
.B Scan the local network for https websites.
 whatweb --no-errors --url-prefix https://192.168.0.0/24
.TP
.B Scan for crossdomain policies in the Alexa Top 1000.
 ./whatweb -i plugin-development/alexa-top-100.txt --url-suffix /crossdomain.xml -p crossdomain_xml


.SH BUGS
Report bugs and feature requests to https://github.com/urbanadventurer/WhatWeb

.SH AUTHOR
Developed by Andrew Horton (urbanadventurer) and Brendan Coles (bcoles).

.SH HOMEPAGE
https://morningstarsecurity.com/research/whatweb

.SH SOURCECODE
https://github.com/urbanadventurer/WhatWeb/
.PP



CHANGELOG
Version 0.5.0 - October 4, 2019

Version 0.5.0 is a major version release form @urbanadventurer and @bcoles. 👏 With the help of the WhatWeb community we have reached over 1800 plugins!  

⚠️ Plugin authors should take note that this release is not backwards compatible, and we have made a migration tool to help you update your private or unreleased plugins.

## FEATURES
* Added WhatWeb Logo and icons
* Added IDN (International Domain Name) support
* Merged 31 new plugins by Bhavin Senjaliya and 22 plugin updates from the unofficial WhatWeb gem at https://rubygems.org/gems/whatweb/
* Added 9 unit tests
* Added loading cookies from a FILE with --cookie-jar=FILE in the document.cookie format
* Added Migration tool to convert plugins to the v0.5 format in plugin-development/migrate-plugins-to-v0.5.rb
* Refactored and modularised the codebase -- @urbanadventurer and @bcoles 

## FIXES
* Bug fix - Ruby 2.4 and 2.5 OpenSSL issue
* Bug fix - Fixed ElasticSearch version 6+ issue
* Bug fix - SQL output issue
* Bug fix - In HTTP Auth parsing @rmaksimov 
* Fix bug causing WhatWeb to exit with no output. Thanks @mguillau42 for tracking down this race condition.
* Typo fix. Thanks Jose Nazario (@paralax)
* Bug fix - clean incorrent UTF-8 byte sequences when loading from file @dirtyfilthy

## MISC
* Updated all plugins with an authors array to show multiple authors
* Plugin list now shows plugin names and websites instead of a truncated plugin description.
* Updated the usage help. Reduced the size of the short usage help.
* Better error reporting in commandline options
* Updated README and converted to Markdown format
* Updated manpage
* Plugins instance variable handling improvement. @Code0x58
* Plugin name defined within plugin as variable. @Code0x58
* Changed from functions to blocks for passive and aggressive plugins. @Code0x58
* Changed plugin locking  @Code0x58
* Changed the behaviour of --grep to only output results that match the grep plugin
* Changed the behaviour of --grep so regexp or text can be matched
* Removed feature --follow-redirect same-domain because it relies on the out-dated TLD library for valid TLDs and SLDs
* Removed requirement for 'nmap' and now use the IPAddr gem for IP address ranges. CIDR, x.x.x.x-x.x.x.x and x.x.x.x-x supported

##  PLUGINS
* 48 new plugins from contributors
* 31 plugins updated from contributors
* 100+ plugin updates by Andrew Horton (@urbanadventurer)

### NEW PLUGINS
* 1n1-hosting.rb - Bhavin Senjaliya (@bhavin1223)
* amazon-elastic-load-balancer.rb - Bhavin Senjaliya (@bhavin1223)
* azure.rb - Bhavin Senjaliya (@bhavin1223)
* bad-behaviour-anti-spam-plugin.rb - Bhavin Senjaliya (@bhavin1223)
* clickmotive.rb - Bhavin Senjaliya (@bhavin1223)
* cdk-connected-website.rb - Bhavin Senjaliya (@bhavin1223)
* craftcms.rb - Bhavin Senjaliya (@bhavin1223)
* f5-big-ip.rb - Bhavin Senjaliya (@bhavin1223)
* hubspot.rb - Bhavin Senjaliya (@bhavin1223)
* jw-player.rb - Bhavin Senjaliya (@bhavin1223)
* laravel.rb - Bhavin Senjaliya (@bhavin1223)
* mezzanine.rb - Bhavin Senjaliya (@bhavin1223)
* moonfruit.rb - Bhavin Senjaliya (@bhavin1223)
* netsuite.rb - Bhavin Senjaliya (@bhavin1223)
* nop-commerce.rb - Bhavin Senjaliya (@bhavin1223)
* pyro-cms.rb - Bhavin Senjaliya (@bhavin1223)
* sailsjs.rb - Bhavin Senjaliya (@bhavin1223)
* schoolwires-centricity.rb - Bhavin Senjaliya (@bhavin1223)
* shopify.rb - Bhavin Senjaliya (@bhavin1223)
* sitecore.rb - Bhavin Senjaliya (@bhavin1223)
* square-space.rb - Bhavin Senjaliya (@bhavin1223)
* tealeaf.rb - Bhavin Senjaliya (@bhavin1223)
* teleflora.rb - Bhavin Senjaliya (@bhavin1223)
* unbounce.rb - Bhavin Senjaliya (@bhavin1223)
* website-tonight.rb - Bhavin Senjaliya (@bhavin1223)
* webtrends.rb - Bhavin Senjaliya (@bhavin1223)
* weebly.rb - Bhavin Senjaliya (@bhavin1223)
* wix.rb - Bhavin Senjaliya (@bhavin1223)
* wordfence.rb - Bhavin Senjaliya (@bhavin1223)
* world-now.rb - Bhavin Senjaliya (@bhavin1223)
* wpml.rb - Bhavin Senjaliya (@bhavin1223)
* bigcommerce.rb - Claudio Salazar (@csalazar)
* demandware.rb - Claudio Salazar (@csalazar)
 *ocp.rb - Claudio Salazar (@csalazar)
* aspforum.rb - Mateusz Golewski (@golewski)
* lithium.rb - Mateusz Golewski (@golewski)
* stackexchange.rb - Mateusz Golewski (@golewski)
* xenforo.rb - Mateusz Golewski (@golewski)
* kinja.rb - Sigit Dewanto
* trbas.rb - Sigit Dewanto
* advance_digitalmg.rb - Elias Dorneles (@eliasdorneles)
* newbay_media.rb - Elias Dorneles (@eliasdorneles)
* gannett.rb - Eugene Amirov (@Allactaga)
* yaf.rb - Eugene Amirov (@Allactaga)
* blox.rb - Shuai Lin (@lins05)
* mcclatchy_interactive.rb - Shuai Lin (@lins05)
* tribune.rb - Shuai Lin (@lins05)
* genexus.rb - Daniel Maldonado (@elcodigok)
* odoo.rb - Naglis Jonaitis (@naglis)

###  PLUGIN UPDATES
* 3dcart.rb - Bhavin Senjaliya (@bhavin1223)
* asp-net.rb - Bhavin Senjaliya (@bhavin1223)
* big-commerce.rb  - Bhavin Senjaliya (@bhavin1223)
* blox.rb - Bhavin Senjaliya (@bhavin1223)
* demandware.rb - Bhavin Senjaliya (@bhavin1223)
* django.rb  - Bhavin Senjaliya (@bhavin1223)
* dot-cms.rb - Bhavin Senjaliya (@bhavin1223)
* DotNetNuke.rb - Bhavin Senjaliya (@bhavin1223)
* drupal.rb - Bhavin Senjaliya (@bhavin1223)
* ektron.rb - Bhavin Senjaliya (@bhavin1223)
* expression-engine.rb  - Bhavin Senjaliya (@bhavin1223)
* incapsula.rb - Bhavin Senjaliya (@bhavin1223)
* kentico.rb - Bhavin Senjaliya (@bhavin1223)
* magento.rb  - Bhavin Senjaliya (@bhavin1223)
* nitix-netscaler.rb - Bhavin Senjaliya (@bhavin1223)
* piwik.rb  - Bhavin Senjaliya (@bhavin1223)
* prestashop.rb  - Bhavin Senjaliya (@bhavin1223)
* ruby-on-rails.rb  - Bhavin Senjaliya (@bhavin1223)
* symfony.rb  - Bhavin Senjaliya (@bhavin1223)
*  blox.rb - Bhavin Senjaliya (@bhavin1223)
* typo3.rb  - Bhavin Senjaliya (@bhavin1223)
* vbulletin.rb - Bhavin Senjaliya (@bhavin1223)
* xenforo.rb - Bhavin Senjaliya (@bhavin1223)
* smf.rb - Claudio Salazar (@csalazar)
* vbulletin.rb - Claudio Salazar (@csalazar)
* bbpress.rb - Shuai Lin (@lins05)

lib/whatweb.rb
# Copyright 2009 to 2020 Andrew Horton and Brendan Coles
#
# This file is part of WhatWeb.
#
# WhatWeb is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# at your option) any later version.
#
# WhatWeb is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with WhatWeb.  If not, see <http://www.gnu.org/licenses/>.

# Debugging
# require 'profile' # debugging

# Standard Ruby
require 'getoptlong'
require 'net/http'
require 'open-uri'
require 'cgi'
require 'thread'
require 'rbconfig' # detect environment, e.g. windows or linux
require 'resolv'
require 'resolv-replace' # asynchronous DNS
require 'open-uri'
require 'digest/md5'
require 'openssl' # required for Ruby version ~> 2.4
require 'socket'
require 'pp'


# WhatWeb libs
require_relative 'whatweb/version.rb'
require_relative 'whatweb/banner.rb'
require_relative 'whatweb/scan.rb'
require_relative 'whatweb/parser.rb'
require_relative 'whatweb/redirect.rb'
require_relative 'gems.rb'
require_relative 'helper.rb'
require_relative 'target.rb'
require_relative 'plugins.rb'
require_relative 'plugin_support.rb'
require_relative 'logging.rb'
require_relative 'colour.rb'
require_relative 'version_class.rb'
require_relative 'http-status.rb'
require_relative 'extend-http.rb'

# load the lib/logging/ folder
Dir["#{File.expand_path(File.dirname(__FILE__))}/logging/*.rb"].each {|file| require file }

# Output options
$WWDEBUG = true # raise exceptions in plugins, etc
$verbose = 0 # $VERBOSE is reserved in ruby
$use_colour = 'auto'
$QUIET = false
$NO_ERRORS = false
$LOG_ERRORS = nil
$PLUGIN_TIMES = Hash.new(0)

# HTTP connection options
$USER_AGENT = "WhatWeb/#{WhatWeb::VERSION}"
$AGGRESSION = 1
$FOLLOW_REDIRECT = 'always'
$USE_PROXY = false
$PROXY_HOST = nil
$PROXY_PORT = 8080
$PROXY_USER = nil
$PROXY_PASS = nil
$HTTP_OPEN_TIMEOUT = 15
$HTTP_READ_TIMEOUT = 30
$WAIT = nil
$CUSTOM_HEADERS = {}
$BASIC_AUTH_USER = nil
$BASIC_AUTH_PASS = nil

$local_host = nil
$local_post = nil

# Ruby Version Compatability
if Gem::Version.new(RUBY_VERSION) < Gem::Version.new(2.0)
  raise('Unsupported version of Ruby. WhatWeb requires Ruby 2.0 or later.')
end

# Initialize HTTP Status class
HTTP_Status.initialize


PLUGIN_DIRS = []

# Load plugins from only one location
# Check for plugins in folders relative to the whatweb file first
# __dir__ follows symlinks
# this will work when whatweb is a symlink in /usr/bin/
$load_path_plugins = [
	File.expand_path('../', __dir__),
	"/usr/share/whatweb" # location Makefile installs into, also used in Kali
]

$load_path_plugins.each do |dir|
	if Dir.exist?(File.expand_path("plugins", dir)) and Dir.exist?(File.expand_path("my-plugins", dir))
		PLUGIN_DIRS << File.expand_path("plugins", dir)
		PLUGIN_DIRS << File.expand_path("my-plugins", dir)
		break
	end
end
